File: src/family_tree/member.rs
use leptos::{RwSignal, SignalGet, SignalGetUntracked, SignalUpdate};
use uuid::Uuid;

#[derive(Debug, Clone, Copy)]
pub struct FamilyMember {
    pub id: Uuid,
    pub name: RwSignal<String>,
    pub is_male: bool,
    pub sons: RwSignal<Vec<FamilyMember>>,
}

impl Default for FamilyMember {
    fn default() -> Self {
        FamilyMember {
            id: Uuid::new_v4(),
            name: RwSignal::new(String::new()),
            is_male: true,
            sons: RwSignal::new(vec![]),
        }
    }
}

impl FamilyMember {
    pub fn new(name: RwSignal<String>) -> Self {
        Self {
            name,
            ..Default::default()
        }
    }
    pub fn with_sons(&mut self, names: &mut Vec<String>) {
        let Some(name) = names.pop() else {
            return;
        };
        let mut son = FamilyMember {
            name: RwSignal::new(name),
            ..Default::default()
        };
        son.with_sons(names);
        self.sons = RwSignal::new(vec![son]);
    }

    pub fn add_son(&self, name: String, is_male: bool) {
        let person = Self::create_from_name(name);
        self.sons.update(|sons| {
            if let Some(son) = sons.iter().find(|x| x.name.get() == person.name.get()) {
                let persons = person.sons.get_untracked().into_iter().collect::<Vec<_>>();
                son.sons.update(|sons| {
                    for person in persons {
                        sons.push(FamilyMember { is_male, ..person })
                    }
                });
            } else {
                sons.push(FamilyMember { is_male, ..person });
            }
        });
    }

    pub fn create_from_name(name: String) -> Self {
        let mut names = name
            .split_whitespace()
            .map(|x| x.to_string())
            .collect::<Vec<_>>();
        let Some(name) = names.pop() else {
            return Self::default();
        };
        let mut person = FamilyMember {
            name: RwSignal::new(name),
            ..Default::default()
        };
        person.with_sons(&mut names);
        person
    }
}

File: ensab_ui/src/app/member.ts
import { signal, WritableSignal } from "@angular/core"

export default class Member {
    id: string;
    name: WritableSignal<String>;
    is_male: boolean;
    sons: WritableSignal<Member[]>;

    constructor(name : string) {
      this.id = crypto.randomUUID();
      this.name = signal(name);
      this.is_male = true;
      this.sons = signal([])
    }

    with_sons(names: string[]): void {
        const name = names.pop(); 
        if(!name){
          return;      
        }
        const son = new Member(name);
        son.with_sons(names);
        this.sons = signal([son]);
    }

    static create_from_name(name: string): Member {
        const names = name.split(' ')
        const namei = names.pop(); 
        if(!namei){
          return new Member("")      
        }
        const person = new Member(namei);
        person.with_sons(names);
        return person;
    }

    add_son(name: string, is_male: boolean) {
        let person_from_name = Member.create_from_name(name);
        this.sons.update(sons => {
            const son_with_similar_name = sons.find(x => x.name() == person_from_name.name());
            if (son_with_similar_name) {
              const similar_sons = son_with_similar_name.sons()
              son_with_similar_name.sons.update(sons => {
                for(const person of similar_sons) {
                  person.is_male = is_male;
                }
                return [...sons,...similar_sons]
              })
              return []
            } else {
              person_from_name.is_male = is_male;
              return [...sons,person_from_name]
            }
        });
    }
  
}
